api/src/main/java/org/openmrs/module/reporting/definition/library/BaseImplementerConfiguredDefinitionLibrary.java

```diff
@@ -0,0 +1,125 @@
+
+	protected static final Log log = LogFactory.getLog(BaseImplementerConfiguredDefinitionLibrary.class);
+
+	private Class<? super T> definitionClass;
+	private String libraryNameSuffix;
+	private File directory;
+
+	private Map<String, T> definitions = new LinkedHashMap<String, T>();
+	private List<LibraryDefinitionSummary> summaries = new ArrayList<LibraryDefinitionSummary>();
+
+	/**
+	 * @param definitionClass the class that all definitions should be a subclass of
+	 * @param libraryNameSuffix what name to give this library (and also subdirectory where definitions are)
+	 */
+	protected BaseImplementerConfiguredDefinitionLibrary(Class<? super T> definitionClass, String libraryNameSuffix,
+	                                                     SerializationService serializationService, File directory) {
+		this.definitionClass = definitionClass;
+		this.libraryNameSuffix = libraryNameSuffix;
+		this.directory = directory;
+
+		OpenmrsSerializer serializer = serializationService.getSerializer(ReportingSerializer.class);
+		for (File file : directory.listFiles()) {
+			String filename = file.getName();
+			String definitionName = filename.substring(0, filename.lastIndexOf('.'));
+			String key = getKeyPrefix() + definitionName;
+			try {
+				log.info("Loading " + file.getAbsolutePath());
+				Definition definition = null;
+				if (filename.endsWith(".reportingserializerxml")) {
+					try {
+						definition = serializer.deserialize(OpenmrsUtil.getFileAsString(file), Definition.class);
+					}
+					catch (SerializationException ex) {
+						log.warn("Invalid serialized definition at " + libraryNameSuffix + " in " + file.getAbsolutePath(), ex);
+					}
+				}
+				else if (filename.endsWith(".sql")) {
+					String sql = OpenmrsUtil.getFileAsString(file);
+					definition = sqlDefinition(sql);
+				}
+				else {
+					log.warn("Don't know how to handle " + file.getAbsolutePath() + " based on file extension");
+				}
+				if (definition != null) {
+					if (definition.getName() == null) {
+						definition.setName(key + ".name");
+					}
+					if (definition.getDescription() == null) {
+						definition.setDescription(key + ".description");
+					}
+					definitions.put(key, (T) definition);
+					summaries.add(summaryFor(key, definition));
+				}
+			}
+			catch (IOException ex) {
+				log.warn("Error reading " + file.getAbsolutePath(), ex);
+			}
+		}
+	}
+
+	@Override
+	public Class<? super T> getDefinitionType() {
+		return definitionClass;
+	}
+
+	@Override
+	public String getKeyPrefix() {
+		return "configuration.definitionlibrary." + libraryNameSuffix + ".";
+	}
+
+	@Override
+	public T getDefinition(String uuid) {
+		return definitions.get(uuid);
+	}
+
+	@Override
+	public List<LibraryDefinitionSummary> getDefinitionSummaries() {
+		return summaries;
+	}
+
+	public boolean hasAnyDefinitions() {
+		return definitions != null && definitions.size() > 0;
+	}
+
+	protected abstract T sqlDefinition(String sql);
```

Comment:
@djazayeri might be nice to do more than this in the base class.  The only thing that differs between one SqlXyzDefinition and another is the name of the class, and _maybe_ the name of the property containing the SQL.  So we might just want to abstract those things out, and then have the base class handle things like interpreting and pulling out parameters, or other common functionality.